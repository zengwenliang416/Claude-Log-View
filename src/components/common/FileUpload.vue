<template>
  <div class="file-upload">
    <input
      ref="fileInput"
      type="file"
      accept=".jsonl,.json"
      @change="handleFileSelect"
      class="file-input"
    />
    
    <button
      class="upload-button"
      :class="{ loading }"
      :disabled="loading"
      @click="triggerFileSelect"
    >
      <LoadingSpinner v-if="loading" size="small" />
      <span v-else class="button-icon">üìÅ</span>
      <span class="button-text">
        {{ loading ? 'Loading...' : 'Load Chat Log' }}
      </span>
    </button>
    
    <!-- Drag and Drop Area -->
    <div
      class="drop-zone"
      :class="{ active: isDragActive, error: dragError }"
      @dragover.prevent="handleDragOver"
      @dragleave.prevent="handleDragLeave"
      @drop.prevent="handleDrop"
      v-if="showDropZone"
    >
      <div class="drop-content">
        <div class="drop-icon">üìé</div>
        <p class="drop-text">
          {{ dragError || 'Drop your .jsonl file here' }}
        </p>
      </div>
    </div>
    
    <!-- File Info -->
    <div class="file-info" v-if="lastLoadedFile">
      <span class="file-name">{{ lastLoadedFile.name }}</span>
      <span class="file-size">{{ formatFileSize(lastLoadedFile.size) }}</span>
    </div>
  </div>
</template>

<script>
import { ref, computed } from 'vue'
import LoadingSpinner from './LoadingSpinner.vue'

export default {
  name: 'FileUpload',
  components: {
    LoadingSpinner
  },
  props: {
    loading: {
      type: Boolean,
      default: false
    }
  },
  emits: ['file-loaded'],
  setup(props, { emit }) {
    const fileInput = ref(null)
    const isDragActive = ref(false)
    const dragError = ref('')
    const lastLoadedFile = ref(null)
    const showDropZone = ref(false)
    
    const triggerFileSelect = () => {
      if (!props.loading && fileInput.value) {
        fileInput.value.click()
      }
    }
    
    const handleFileSelect = (event) => {
      const file = event.target.files[0]
      if (file) {
        processFile(file)
      }
    }
    
    const handleDragOver = (event) => {
      event.preventDefault()
      isDragActive.value = true
      dragError.value = ''
      
      // Validate file type during drag
      const items = event.dataTransfer.items
      if (items && items.length > 0) {
        const item = items[0]
        if (item.kind === 'file') {
          const file = item.getAsFile()
          if (file && !isValidFileType(file)) {
            dragError.value = 'Invalid file type. Please use .jsonl or .json files.'
          }
        }
      }
    }
    
    const handleDragLeave = (event) => {
      event.preventDefault()
      if (!event.relatedTarget || !event.currentTarget.contains(event.relatedTarget)) {
        isDragActive.value = false
        dragError.value = ''
      }
    }
    
    const handleDrop = (event) => {
      event.preventDefault()
      isDragActive.value = false
      dragError.value = ''
      
      const files = event.dataTransfer.files
      if (files.length > 0) {
        const file = files[0]
        if (isValidFileType(file)) {
          processFile(file)
        } else {
          dragError.value = 'Invalid file type. Please use .jsonl or .json files.'
          setTimeout(() => {
            dragError.value = ''
          }, 3000)
        }
      }
    }
    
    const isValidFileType = (file) => {
      return file.name.endsWith('.jsonl') || file.name.endsWith('.json')
    }
    
    const processFile = (file) => {
      if (!isValidFileType(file)) {
        alert('Please select a .jsonl or .json file.')
        return
      }
      
      // Check file size (limit to 50MB)
      const maxSize = 50 * 1024 * 1024 // 50MB
      if (file.size > maxSize) {
        alert(`File too large. Maximum size is ${Math.round(maxSize / (1024 * 1024))}MB.`)
        return
      }
      
      lastLoadedFile.value = {
        name: file.name,
        size: file.size,
        lastModified: file.lastModified
      }
      
      emit('file-loaded', file)
      
      // Reset file input
      if (fileInput.value) {
        fileInput.value.value = ''
      }
    }
    
    const formatFileSize = (bytes) => {
      if (bytes === 0) return '0 Bytes'
      
      const k = 1024
      const sizes = ['Bytes', 'KB', 'MB', 'GB']
      const i = Math.floor(Math.log(bytes) / Math.log(k))
      
      return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i]
    }
    
    // Show drop zone on drag enter to window
    const handleWindowDragEnter = (event) => {
      if (event.dataTransfer.types.includes('Files')) {
        showDropZone.value = true
      }
    }
    
    const handleWindowDragLeave = (event) => {
      // Hide drop zone when leaving window
      if (event.clientX === 0 && event.clientY === 0) {
        showDropZone.value = false
      }
    }
    
    // Add window drag events
    if (typeof window !== 'undefined') {
      window.addEventListener('dragenter', handleWindowDragEnter)
      window.addEventListener('dragleave', handleWindowDragLeave)
    }
    
    return {
      fileInput,
      isDragActive,
      dragError,
      lastLoadedFile,
      showDropZone,
      triggerFileSelect,
      handleFileSelect,
      handleDragOver,
      handleDragLeave,
      handleDrop,
      formatFileSize
    }
  }
}
</script>

<style scoped>
.file-upload {
  position: relative;
}

.file-input {
  display: none;
}

.upload-button {
  display: flex;
  align-items: center;
  gap: var(--spacing-sm);
  padding: var(--spacing-sm) var(--spacing-md);
  background-color: var(--accent-color);
  color: white;
  border: none;
  border-radius: var(--border-radius);
  font-size: var(--font-size-sm);
  font-weight: var(--font-weight-medium);
  cursor: pointer;
  transition: all var(--transition-fast);
}

.upload-button:hover:not(:disabled) {
  background-color: var(--accent-hover);
  transform: translateY(-1px);
}

.upload-button:disabled {
  opacity: 0.7;
  cursor: not-allowed;
  transform: none;
}

.upload-button.loading {
  background-color: var(--bg-tertiary);
  color: var(--text-secondary);
}

.button-icon {
  font-size: var(--font-size-base);
}

.button-text {
  white-space: nowrap;
}

.drop-zone {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.8);
  backdrop-filter: blur(4px);
  z-index: var(--z-modal);
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0;
  visibility: hidden;
  transition: all var(--transition-normal);
}

.drop-zone.active {
  opacity: 1;
  visibility: visible;
}

.drop-zone.error {
  background-color: rgba(255, 107, 107, 0.1);
}

.drop-content {
  text-align: center;
  padding: var(--spacing-xl);
  background-color: var(--bg-secondary);
  border: 2px dashed var(--border-color);
  border-radius: var(--border-radius-lg);
  transition: all var(--transition-fast);
}

.drop-zone.active .drop-content {
  border-color: var(--accent-color);
  background-color: var(--bg-tertiary);
}

.drop-zone.error .drop-content {
  border-color: var(--error-color);
  background-color: rgba(255, 107, 107, 0.1);
}

.drop-icon {
  font-size: 3rem;
  margin-bottom: var(--spacing-md);
  opacity: 0.6;
}

.drop-text {
  margin: 0;
  font-size: var(--font-size-lg);
  color: var(--text-primary);
  font-weight: var(--font-weight-medium);
}

.drop-zone.error .drop-text {
  color: var(--error-color);
}

.file-info {
  display: flex;
  flex-direction: column;
  margin-top: var(--spacing-sm);
  font-size: var(--font-size-xs);
  color: var(--text-muted);
}

.file-name {
  font-weight: var(--font-weight-medium);
  color: var(--text-secondary);
}

.file-size {
  margin-top: 2px;
}

/* Responsive design */
@media (max-width: 768px) {
  .upload-button {
    padding: var(--spacing-xs) var(--spacing-sm);
    font-size: var(--font-size-xs);
  }
  
  .button-text {
    display: none;
  }
  
  .drop-content {
    padding: var(--spacing-lg);
    margin: var(--spacing-md);
  }
  
  .drop-icon {
    font-size: 2rem;
  }
  
  .drop-text {
    font-size: var(--font-size-base);
  }
}

@media (max-width: 480px) {
  .upload-button {
    padding: var(--spacing-xs);
  }
  
  .button-icon {
    font-size: var(--font-size-sm);
  }
}
</style>