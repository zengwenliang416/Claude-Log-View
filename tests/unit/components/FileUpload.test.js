import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { nextTick } from 'vue'
import FileUpload from '@/components/common/FileUpload.vue'

describe('FileUpload.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(FileUpload, {
      props: { loading: false }
    })
  })

  afterEach(() => {
    wrapper.unmount()
  })

  describe('rendering', () => {
    it('should render upload button', () => {
      const button = wrapper.find('.upload-button')
      expect(button.exists()).toBe(true)
      expect(button.text()).toContain('Load Chat Log')
    })

    it('should hide file input', () => {
      const input = wrapper.find('.file-input')
      expect(input.exists()).toBe(true)
      expect(input.element.style.display).toBe('none')
    })

    it('should not show drop zone initially', () => {
      const dropZone = wrapper.find('.drop-zone')
      expect(dropZone.exists()).toBe(false)
    })

    it('should show correct button icon when not loading', () => {
      const icon = wrapper.find('.button-icon')
      expect(icon.text()).toBe('📁')
    })

    it('should show loading spinner when loading', async () => {
      await wrapper.setProps({ loading: true })
      
      const spinner = wrapper.findComponent({ name: 'LoadingSpinner' })
      expect(spinner.exists()).toBe(true)
      
      const buttonText = wrapper.find('.button-text')
      expect(buttonText.text()).toBe('Loading...')
    })
  })

  describe('file selection', () => {
    it('should trigger file input click when button is clicked', async () => {
      const fileInput = wrapper.find('.file-input')
      const clickSpy = vi.spyOn(fileInput.element, 'click')
      
      const button = wrapper.find('.upload-button')
      await button.trigger('click')
      
      expect(clickSpy).toHaveBeenCalled()
    })

    it('should not trigger file input when loading', async () => {
      await wrapper.setProps({ loading: true })
      
      const fileInput = wrapper.find('.file-input')
      const clickSpy = vi.spyOn(fileInput.element, 'click')
      
      const button = wrapper.find('.upload-button')
      await button.trigger('click')
      
      expect(clickSpy).not.toHaveBeenCalled()
    })

    it('should emit file-loaded event when valid file is selected', async () => {
      const file = createMockFile('{"test": "data"}', 'test.jsonl')
      const fileInput = wrapper.find('.file-input')
      
      // Mock the files property
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      await nextTick()
      
      expect(wrapper.emitted('file-loaded')).toBeTruthy()
      expect(wrapper.emitted('file-loaded')[0][0]).toBe(file)
    })

    it('should show file info after successful upload', async () => {
      const file = createMockFile('{"test": "data"}', 'test.jsonl')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      await nextTick()
      
      const fileInfo = wrapper.find('.file-info')
      expect(fileInfo.exists()).toBe(true)
      
      const fileName = wrapper.find('.file-name')
      expect(fileName.text()).toBe('test.jsonl')
      
      const fileSize = wrapper.find('.file-size')
      expect(fileSize.text()).toContain('Bytes')
    })

    it('should reset file input after processing', async () => {
      const file = createMockFile('{"test": "data"}', 'test.jsonl')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      await nextTick()
      
      expect(fileInput.element.value).toBe('')
    })
  })

  describe('file validation', () => {
    it('should accept .jsonl files', async () => {
      const file = createMockFile('{"test": "data"}', 'test.jsonl')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      
      expect(wrapper.emitted('file-loaded')).toBeTruthy()
    })

    it('should accept .json files', async () => {
      const file = createMockFile('{"test": "data"}', 'test.json')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      
      expect(wrapper.emitted('file-loaded')).toBeTruthy()
    })

    it('should reject invalid file types', async () => {
      const alertSpy = vi.spyOn(window, 'alert').mockImplementation(() => {})
      const file = createMockFile('content', 'test.txt')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      
      expect(alertSpy).toHaveBeenCalledWith('Please select a .jsonl or .json file.')
      expect(wrapper.emitted('file-loaded')).toBeFalsy()
      
      alertSpy.mockRestore()
    })

    it('should reject files that are too large', async () => {
      const alertSpy = vi.spyOn(window, 'alert').mockImplementation(() => {})
      const largeContent = 'x'.repeat(51 * 1024 * 1024) // 51MB
      const file = createMockFile(largeContent, 'large.jsonl')
      const fileInput = wrapper.find('.file-input')
      
      Object.defineProperty(fileInput.element, 'files', {
        value: [file],
        writable: false
      })
      
      await fileInput.trigger('change')
      
      expect(alertSpy).toHaveBeenCalledWith('File too large. Maximum size is 50MB.')
      expect(wrapper.emitted('file-loaded')).toBeFalsy()
      
      alertSpy.mockRestore()
    })
  })

  describe('drag and drop', () => {
    beforeEach(async () => {
      // Simulate showing drop zone
      await wrapper.setData({ showDropZone: true })
    })

    it('should show drop zone when dragging files over window', async () => {
      await wrapper.setData({ showDropZone: true })
      
      const dropZone = wrapper.find('.drop-zone')
      expect(dropZone.exists()).toBe(true)
    })

    it('should activate drop zone on drag over', async () => {
      await wrapper.setData({ showDropZone: true })
      
      const dropZone = wrapper.find('.drop-zone')
      const mockEvent = {
        preventDefault: vi.fn(),
        dataTransfer: {
          items: [{
            kind: 'file',
            getAsFile: () => createMockFile('content', 'test.jsonl')
          }]
        }
      }
      
      await dropZone.trigger('dragover', mockEvent)
      
      expect(dropZone.classes()).toContain('active')
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })

    it('should show error for invalid file type during drag', async () => {
      await wrapper.setData({ showDropZone: true })
      
      const dropZone = wrapper.find('.drop-zone')
      const mockEvent = {
        preventDefault: vi.fn(),
        dataTransfer: {
          items: [{
            kind: 'file',
            getAsFile: () => createMockFile('content', 'test.txt')
          }]
        }
      }
      
      await dropZone.trigger('dragover', mockEvent)
      await nextTick()
      
      const dropText = wrapper.find('.drop-text')
      expect(dropText.text()).toContain('Invalid file type')
    })

    it('should deactivate drop zone on drag leave', async () => {
      await wrapper.setData({ showDropZone: true, isDragActive: true })
      
      const dropZone = wrapper.find('.drop-zone')
      const mockEvent = {
        preventDefault: vi.fn(),
        relatedTarget: null,
        currentTarget: { contains: () => false }
      }
      
      await dropZone.trigger('dragleave', mockEvent)
      
      expect(wrapper.vm.isDragActive).toBe(false)
    })

    it('should process valid dropped files', async () => {
      await wrapper.setData({ showDropZone: true })
      
      const dropZone = wrapper.find('.drop-zone')
      const file = createMockFile('{"test": "data"}', 'test.jsonl')
      const mockEvent = {
        preventDefault: vi.fn(),
        dataTransfer: {
          files: [file]
        }
      }
      
      await dropZone.trigger('drop', mockEvent)
      await nextTick()
      
      expect(wrapper.emitted('file-loaded')).toBeTruthy()
      expect(wrapper.emitted('file-loaded')[0][0]).toBe(file)
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })

    it('should show error for invalid dropped files', async () => {
      await wrapper.setData({ showDropZone: true })
      
      const dropZone = wrapper.find('.drop-zone')
      const file = createMockFile('content', 'test.txt')
      const mockEvent = {
        preventDefault: vi.fn(),
        dataTransfer: {
          files: [file]
        }
      }
      
      await dropZone.trigger('drop', mockEvent)
      await nextTick()
      
      expect(wrapper.vm.dragError).toContain('Invalid file type')
      expect(wrapper.emitted('file-loaded')).toBeFalsy()
    })
  })

  describe('utility functions', () => {
    it('should format file sizes correctly', () => {
      const { formatFileSize } = wrapper.vm
      
      expect(formatFileSize(0)).toBe('0 Bytes')
      expect(formatFileSize(1024)).toBe('1 KB')
      expect(formatFileSize(1024 * 1024)).toBe('1 MB')
      expect(formatFileSize(1536)).toBe('1.5 KB')
    })

    it('should validate file types correctly', () => {
      const validJsonl = createMockFile('content', 'test.jsonl')
      const validJson = createMockFile('content', 'test.json')
      const invalid = createMockFile('content', 'test.txt')
      
      expect(wrapper.vm.isValidFileType(validJsonl)).toBe(true)
      expect(wrapper.vm.isValidFileType(validJson)).toBe(true)
      expect(wrapper.vm.isValidFileType(invalid)).toBe(false)
    })
  })

  describe('accessibility', () => {
    it('should have proper ARIA attributes', () => {
      const button = wrapper.find('.upload-button')
      expect(button.attributes('role')).toBe('button')
    })

    it('should be keyboard accessible', async () => {
      const button = wrapper.find('.upload-button')
      const fileInput = wrapper.find('.file-input')
      const clickSpy = vi.spyOn(fileInput.element, 'click')
      
      // Simulate Enter key press
      await button.trigger('keydown.enter')
      
      expect(clickSpy).toHaveBeenCalled()
    })

    it('should have proper disabled state', async () => {
      await wrapper.setProps({ loading: true })
      
      const button = wrapper.find('.upload-button')
      expect(button.attributes('disabled')).toBeDefined()
      expect(button.classes()).toContain('loading')
    })
  })

  describe('responsive behavior', () => {
    it('should adapt to mobile viewports', () => {
      // This would typically require viewport mocking or CSS testing
      const button = wrapper.find('.upload-button')
      expect(button.exists()).toBe(true)
      
      // Test that mobile-specific classes are properly applied
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})