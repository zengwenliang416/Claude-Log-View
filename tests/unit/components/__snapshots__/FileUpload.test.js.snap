// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileUpload.vue > responsive behavior > should adapt to mobile viewports 1`] = `
"<div data-v-a1ed4b7d="" class="file-upload"><input data-v-a1ed4b7d="" type="file" accept=".jsonl,.json" class="file-input"><button data-v-a1ed4b7d="" class="upload-button"><span data-v-a1ed4b7d="" class="button-icon">📁</span><span data-v-a1ed4b7d="" class="button-text">Load Chat Log</span></button><!-- Drag and Drop Area -->
  <!--v-if-->
  <!-- File Info -->
  <!--v-if-->
</div>"
`;
